iroh-bytes/src/util/progress.rs

```diff
@@ -391,3 +393,112 @@ impl From<ProgressSendError> for std::io::Error {
         std::io::Error::new(std::io::ErrorKind::BrokenPipe, e)
     }
 }
+
+/// A slice writer that adds a synchronous progress callback.
+///
+/// This wraps any `AsyncSliceWriter`, passes through all operations to the inner writer, and
+/// calls the passed `on_write` callback whenever data is written.
+#[derive(Debug)]
+pub struct ProgressSliceWriter<W, F>(W, F);
+
+impl<W: AsyncSliceWriter, F: FnMut(u64)> ProgressSliceWriter<W, F> {
+    /// Create a new `ProgressSliceWriter` from an inner writer and a progress callback
+    ///
+    /// The `on_write` function is called for each write, with the `offset` as the first and the
+    /// length of the data as the second param.
+    pub fn new(inner: W, on_write: F) -> Self {
+        Self(inner, on_write)
+    }
+
+    /// Return the inner writer
+    pub fn into_inner(self) -> W {
+        self.0
+    }
+}
+
+impl<W: AsyncSliceWriter + 'static, F: FnMut(u64, usize) + 'static> AsyncSliceWriter
+    for ProgressSliceWriter<W, F>
+{
+    type WriteBytesAtFuture<'a> = W::WriteBytesAtFuture<'a>;
+    fn write_bytes_at(&mut self, offset: u64, data: Bytes) -> Self::WriteBytesAtFuture<'_> {
+        (self.1)(offset, data.len());
+        self.0.write_bytes_at(offset, data)
+    }
+
+    type WriteAtFuture<'a> = W::WriteAtFuture<'a>;
+    fn write_at<'a>(&'a mut self, offset: u64, data: &'a [u8]) -> Self::WriteAtFuture<'a> {
+        (self.1)(offset, data.len());
+        self.0.write_at(offset, data)
+    }
+
+    type SyncFuture<'a> = W::SyncFuture<'a>;
+    fn sync(&mut self) -> Self::SyncFuture<'_> {
+        self.0.sync()
+    }
+
+    type SetLenFuture<'a> = W::SetLenFuture<'a>;
+    fn set_len(&mut self, size: u64) -> Self::SetLenFuture<'_> {
+        self.0.set_len(size)
+    }
+}
+
+/// A slice writer that adds an asynchronous progress callback.
+///
+/// This wraps any `AsyncSliceWriter`, passes through all operations to the inner writer, and
+/// calls the passed `on_write` callback whenever data is written. `on_write` must return a future
+/// which is awaited before continuing with the write.
+#[derive(Debug)]
+pub struct FallibleProgressSliceWriter<W, F>(W, F);
+
+impl<W: AsyncSliceWriter, F: Fn(u64, usize) -> io::Result<()> + 'static>
```

Comment:
The description above says that F is an async fn. But it is just a sync, but fallible fn. So I think either the description or the code is wrong.

Reply:
Very correct ;) pushed a fix. wrote docs before rereading the code, and then forgot to update.